core/src/main/java/org/apache/accumulo/core/metadata/TabletLocationState.java

```diff
@@ -119,15 +120,23 @@ public boolean hasSuspend() {
 
   public TabletState getState(Set<TServerInstance> liveServers) {
     if (hasFuture()) {
-      return liveServers.contains(future) ? TabletState.ASSIGNED
+      return liveServers.contains(future.getServerInstance()) ? TabletState.ASSIGNED
           : TabletState.ASSIGNED_TO_DEAD_SERVER;
     } else if (hasCurrent()) {
-      return liveServers.contains(current) ? TabletState.HOSTED
+      return liveServers.contains(current.getServerInstance()) ? TabletState.HOSTED
           : TabletState.ASSIGNED_TO_DEAD_SERVER;
     } else if (hasSuspend()) {
       return TabletState.SUSPENDED;
     } else {
       return TabletState.UNASSIGNED;
     }
   }
+
+  private static TabletMetadata.Location validateLocation(TabletMetadata.Location location,
+      TabletMetadata.LocationType type) {
+    if (location != null && !location.getType().equals(type)) {
+      throw new IllegalArgumentException("Location type is required to be of type " + type);
+    }
+    return location;
+  }
```

Comment:
Wondering if the future,current, and last instance var should keep the type TServerInstance.  It depends on what the most frequent use is, which I am not sure of.  I saw one class while looking over this that was always converting these fields to TserverInstance. If it were changed back, could still use this nice validation method with a slight change.

```suggestion
  private static TserverInstance validateLocation(TabletMetadata.Location location,
      TabletMetadata.LocationType type) {
    if (location != null && !location.getType().equals(type)) {
      throw new IllegalArgumentException("Location type is required to be of type " + type);
    }
    return location.getServerInstance();
  }
```

Reply:
Looking through the PR it seems like TabletGroupWatcher would be the main beneficiary of keep the current type, but not sure what class would not benefit.

Reply:
There are two constructors, would this mean 6 static methods are needed? Six static methods seems fine.  Making the constructors private if it makes the code shorter and more consistent SGTM.